src/table/use-sticky-columns.tsx

```diff
@@ -0,0 +1,224 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useLayoutEffect, useState, createRef, useEffect, useCallback, useMemo } from 'react';
+import { TableProps } from './interfaces';
+import { CellOffsets } from './internal';
+import clsx from 'clsx';
+interface StickyStyles {
+  sticky: {
+    left?: string;
+    right?: string;
+  };
+  stuck: {
+    paddingLeft?: string;
+  };
+}
+
+interface StickyColumnParams {
+  containerWidth: number | null;
+  hasSelection: boolean;
+  isWrapperScrollable: boolean;
+  stickyColumns?: TableProps.StickyColumns;
+  visibleColumnsLength: number;
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}
+
+export interface GetStickyColumnProperties {
+  isSticky: boolean;
+  isLastStickyLeft: boolean;
+  isLastStickyRight: boolean;
+  stickyStyles: StickyStyles;
+}
+
+// We allow the table to have a minimum of 148px of available space besides the sum of the widths of the sticky columns
+// This value is an UX recommendation and is approximately 1/3 of our smallest breakpoint (465px)
+const MINIMUM_SCROLLABLE_SPACE = 148;
+
+const useCellOffsets = (tableCellRefs: Array<React.RefObject<HTMLTableCellElement>>) => {
+  // This hook calculates the cumulative offsets of the cells in each column of the table, based on the widths of their siblings
+
+  const [cellOffsets, setCellOffsets] = useState<CellOffsets>({ first: [], last: [] });
+
+  const updateCellOffsets = useCallback(() => {
+    // Calculate widths of all previous siblings of each table cell in the `tableCellRefs` array
+    let firstColumnsOffsets = tableCellRefs
+      .map(ref => (ref?.current?.previousSibling as HTMLTableCellElement)?.offsetWidth)
+      .filter(Boolean);
```

Comment:
If I understood the idea correctly, the `firstColumnsOffsets` represents the columns cumulative widths from the start and it referenced by index, like: `firstColumnsOffsets[1]` - the widths of 1 and 2 columns. If that's the case, the `.filter(Boolean)` will break the indexing. I don't think that is possible though because it is unlikely for any columns to be of width 0.

I propose refactoring the code for making the intent clear (provided I got it right):
```
const firstColumnsOffsets = []

for (let i = 0; i < tableCellRefs.length; i++) {
  const ref = tableCellRefs[i];
  if (!ref || !ref.current) {
    throw new Error('Invariant violation: table cell reference is not set.');
  }
  const width = ref.current.previousSibling as HTMLTableCellElement)?.offsetWidth ?? 0;
  firstColumnsOffsets[i] = i == 0 ? width : firstColumnsOffsets[i - 1] + width;
}
```